### PR TITLE
refactor southbound templates api

### PR DESCRIPTION
**Pull request by lanchongyizu**
## Background

Currently, RackHD is heavily coupled with and dependent on lookups, especially the 1 to 1 relationship between MAC and IP, which causes the static IP problems and maybe other problems.
There are two kinds of relationships between MAC and IP, which should be taken into consideration now.
1) one MAC address could map to multiple IP addresses, which exists in ARP cache.
2) one IP address could map to multiple MAC addresses, when users want to set multiple nodes with the same IP address, and it may exist in ARP cache sometime.

When the nodeId assigned, we can use it to identify the node rather than depend on lookups, since the latter introduces more complexity into lookups.
Also, MACAddress and maybe serial board number can be as identifiers.

## Implementation

Thereafter, the PRs

> https://github.com/RackHD/on-http/pull/531 (this PR)
> https://github.com/RackHD/on-tasks/pull/374
> https://github.com/RackHD/on-taskgraph/pull/187
> https://github.com/RackHD/on-skupack/pull/57
> https://github.com/RackHD/docs/pull/362

 are to refactor

`/templates/:name`

to 

```
/templates/:name?nodeId=:nodeId
/templates/:name?macs=:macs
```

**Additional work:**

- make option downloadUrl to use FULL PATH
- remove postLookup() in esx-ks

## Test

I have tested the PRs by successfully installing below OSes which would be functionally equal to Regression Test.
**Note: For Windows Server 2012, you should regenerate a new WinPE image.**
```
Ubuntu 14.04
ESXi 6.0
CentOS 6.5
CentOS 7.0
openSUSE 42.1
CoreOS
Windows Server 2012
Photon OS
```

@RackHD/corecommitters @panpan0000 @iceiilin @keedya @johren

### CODE REVIEW

**Comment by lanchongyizu** [comment]
jenkins: depends on https://github.com/RackHD/on-tasks/pull/374

**Comment by changev** [comment]
test this please

**Comment by changev** [comment]
test this please

**Comment by lanchongyizu** [comment]
test this please.

**Comment by changev** [comment]
test this please

**Comment by brianparry** [comment]
@keedya Would you mind taking a look through this?  I know you had some similar ideas about using nodeId in place of a lookup.

**Comment by brianparry** [comment]
@lanchongyizu  Changes look good, but we need to make sure that functional and regression tests pass before merging.  Please work with @anhou and @yyscamper to coordinate the merge.

**Comment by keedya** [comment]
looks good :+1:

**Comment by lanchongyizu** [comment]
test this please.

**Comment by lanchongyizu** [comment]
test this please.

**Comment by lanchongyizu** [comment]
Currently, the Jenkins-Dependency-PR can't work properly.

**Comment by lanchongyizu** [comment]
test this please.

**Comment by lanchongyizu** [comment]
test this please.

**Comment by lanchongyizu** [comment]
test this please.

File: spec/lib/api/2.0/templates-spec.js [hunk 1]
```diff
@@ -177,15 +177,10 @@ describe('Http.Api.Templates', function () {
...
-            waterline.nodes = {
-                needByIdentifier: sinon.stub().resolves({
-                    id: 'node id'
-                })
-            };
             findActiveGraphForTarget.resolves(graph);
             this.sandbox.stub(swagger, 'makeRenderableOptions').resolves({});
             taskProtocol.requestProperties.resolves({});
-            return helper.request().get('/api/2.0/templates/123')
+            return helper.request().get('/api/2.0/templates/123?nodeId=123')
```
Comment (lanchongyizu) [inline spec/lib/api/2.0/templates-spec.js]: Done, thanks.

File: lib/api/2.0/templates.js [hunk 2]
```diff
@@ -26,23 +26,15 @@ var templatesMetaGetByName = controller(function(req) {
...
-        return waterline.nodes.needByIdentifier(mac);
-    })
-    .then(function(node) {
-        return Promise.all([
-            workflowApiService.findActiveGraphForTarget(node.id),
-            node
-        ]);
-    })
-    .spread(function(workflow, node) {
+    var nodeId = req.swagger.query.nodeId;
```
Comment (lanchongyizu) [inline lib/api/2.0/templates.js]: We have created story https://www.pivotaltracker.com/story/show/134938873 to follow the suggestions.
Thanks.

**Comment by lanchongyizu** [comment]
@brianparry OK, thanks for your review.

**Comment by lanchongyizu** [comment]
test this please.

**Comment by leoyjchang** [comment]
@lanchongyizu Very appreciate you updated the PR description following the template, which makes people to understand what been done clearly.  Awesome~~

**Comment by lanchongyizu** [comment]
test this please.

**Pull request merged**